File: clients/python/test/grpc_smoke.py
"""Runtime smoke: drive the gRPC transport against a live `pawrly serve`.

Exercises the gRPC-only guarantees the REST smoke can't: a non-empty server
query id, lossless typed Arrow values, and a real `shutdown`.
"""

import os
import sys
import time

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "src"))

import pawrly  # noqa: E402


def fail(msg: str) -> None:
    print("GRPC SMOKE FAIL:", msg)
    sys.exit(1)


endpoint = os.environ.get("PAWRLY_GRPC")
if not endpoint:
    fail("PAWRLY_GRPC not set (e.g. tcp://127.0.0.1:8788)")

c = pawrly.PawrlyClient.grpc(endpoint)

deadline = time.time() + 10
while True:
    try:
        if c.health().ok:
            break
    except Exception:
        pass
    if time.time() > deadline:
        fail("serve never came up")
    time.sleep(0.1)

h = c.health()
print("health:", h)
if not h.ok:
    fail("health not ok")

# query() streams Arrow IPC frames; values stay typed (n is an int, not "1").
handle = c.query("SELECT 1 AS n")
if not handle.id:
    fail("gRPC query handle has no server id (Started frame missing)")
print("query id:", handle.id)
res = handle.collect()
print("query:", res)
if res.row_count != 1 or res.rows[0].get("n") != 1:
    fail(f"unexpected query result {res}")
if not isinstance(res.rows[0]["n"], int):
    fail(f"expected typed int, got {type(res.rows[0]['n'])}")

# streaming iteration (Arrow batches under the hood).
streamed = list(c.query("SELECT 1 AS n"))
if len(streamed) != 1:
    fail(f"expected 1 streamed row, got {len(streamed)}")
print("streamed rows:", len(streamed))

# explain returns a plan string.
plan = c.explain("SELECT 1 AS n", False)
if "1" not in plan:
    fail(f"explain plan looks wrong: {plan!r}")
print("explain ok (", len(plan), "chars )")

# materialize a query result, then read it back.
out = c.materialize("grpc_mat_py", pawrly.MaterializeSpec(sql="SELECT 1 AS n"))
print("materialize:", out)
if out.name.get("table") != "grpc_mat_py":
    fail(f"materialize name {out.name}")
back = c.query("SELECT * FROM materialized.grpc_mat_py").collect()
if back.row_count != 1:
    fail(f"materialized readback row_count {back.row_count}")
print("materialized readback row_count:", back.row_count)

# cancel round-trips a bool; an already-finished id is simply False (no raise).
cancelled = c.cancel("does-not-exist")
if not isinstance(cancelled, bool):
    fail(f"cancel returned {cancelled!r}, expected bool")
print("cancel(unknown) ->", cancelled)

# semantic_query on an empty workspace (no models) surfaces a PawrlyError.
try:
    c.semantic_query(pawrly.SemanticQuery(measures=["orders.revenue"])).collect()
    fail("semantic_query should have errored (no models)")
except pawrly.PawrlyError as e:
    print("semantic_query -> PawrlyError:", e.code)

# shutdown is real over gRPC (do it last — it stops the server).
c.shutdown()
print("shutdown -> ok")

print("GRPC SMOKE OK")

File: clients/typescript/test/grpc_smoke.mjs
// Runtime smoke: drive the built gRPC transport against a live `pawrly serve`.
// Exercises the gRPC-only guarantees the REST smoke can't: a non-empty server
// query id, lossless typed Arrow values, and a real `shutdown`.
import { PawrlyClient, PawrlyError } from "../dist/index.js";

const endpoint = process.env.PAWRLY_GRPC;
if (!endpoint) {
  console.error("PAWRLY_GRPC not set (e.g. tcp://127.0.0.1:8788)");
  process.exit(1);
}

const client = new PawrlyClient({ transport: "grpc", endpoint });

function fail(msg) {
  console.error("GRPC SMOKE FAIL:", msg);
  process.exit(1);
}

// BigInt-safe stringify (64-bit ints beyond 2^53 stay BigInt).
const j = (x) => JSON.stringify(x, (_, v) => (typeof v === "bigint" ? v.toString() : v));

// Wait for serve to come up.
const deadline = Date.now() + 10_000;
for (;;) {
  try {
    if ((await client.health()).ok) break;
  } catch {
    /* not up yet */
  }
  if (Date.now() > deadline) fail("serve never came up");
  await new Promise((r) => setTimeout(r, 100));
}

const h = await client.health();
console.log("health:", JSON.stringify(h));
if (!h.ok) fail("health not ok");

// query() streams Arrow IPC frames; values stay typed, and the handle carries
// the server-assigned query id (empty over REST).
const handle = await client.query("SELECT 1 AS n");
if (!handle.id) fail("gRPC query handle has no server id (Started frame missing)");
console.log("query id:", handle.id);
const res = await handle.collect();
console.log("query:", j(res));
if (res.rowCount !== 1) fail(`expected rowCount 1, got ${res.rowCount}`);
if (res.rows[0]?.n !== 1) fail(`expected n=1, got ${JSON.stringify(res.rows[0])}`);
if (typeof res.rows[0].n !== "number") {
  fail(`expected typed number, got ${typeof res.rows[0].n}`);
}

// Async iteration (Arrow batches under the hood).
let streamed = 0;
for await (const row of await client.query("SELECT 1 AS n")) {
  streamed += 1;
  if (row.n !== 1) fail(`stream row n=${JSON.stringify(row)}`);
}
if (streamed !== 1) fail(`expected 1 streamed row, got ${streamed}`);
console.log("streamed rows:", streamed);

// explain returns a plan string.
const plan = await client.explain("SELECT 1 AS n", false);
if (!plan.includes("1")) fail(`explain plan looks wrong: ${plan}`);
console.log("explain ok (", plan.length, "chars )");

// materialize a query result, then read it back.
const out = await client.materialize("grpc_mat_ts", { kind: "query", sql: "SELECT 1 AS n" });
console.log("materialize:", j(out));
if (out.name.table !== "grpc_mat_ts") fail(`materialize name ${JSON.stringify(out.name)}`);
const back = await (await client.query("SELECT * FROM materialized.grpc_mat_ts")).collect();
if (back.rowCount !== 1) fail(`materialized readback rowCount ${back.rowCount}`);
console.log("materialized readback rowCount:", back.rowCount);

// cancel round-trips a bool; an already-finished id is simply false (no throw).
const cancelled = await client.cancel("does-not-exist");
if (typeof cancelled !== "boolean") fail(`cancel returned ${cancelled}, expected boolean`);
console.log("cancel(unknown) ->", cancelled);

// semanticQuery on an empty workspace (no models) surfaces a PawrlyError — the
// status-level error path (mapped from a Connect status, not an in-stream frame).
try {
  await client.semanticQuery({ measures: ["orders.revenue"] });
  fail("semanticQuery should have errored (no models)");
} catch (e) {
  if (!(e instanceof PawrlyError)) fail(`semanticQuery wrong error: ${e}`);
  console.log("semanticQuery -> PawrlyError:", e.code);
}

// shutdown is real over gRPC (do it last — it stops the server).
await client.shutdown();
console.log("shutdown -> ok");

client.close();
console.log("GRPC SMOKE OK");
